### ISSUE TITLE
Improve GAN Network 

### ISSUE THREAD

**Issue by Skylion007** [ppwwyyxx/tensorpack#105]
Hello, let me just say I am a big fan of the code in this repo. I noticed though that the reference implementation you use for the GAN example is from November of 2015. I'd like to suggest that the code be improved by utilizing the latest techniques from a recent OpenAI [paper](https://arxiv.org/pdf/1606.03498v1.pdf). The code can be found [here](https://github.com/openai/improved-gan) and is already in Tensorflow, so it shouldn't be too hard to import. It would be a welcome addition since the main reason I use this project is for the very efficient GAN implementation. From what I've read from skimming over the paper, OpenAI modifies the loss function for the generator and discriminator to make them converge faster.

**Comment by PatWie** [comment]
I have implemented these changes in my version:
- mini-batch discrimination
- switched the labels 0 <-> 1
- label-smoothing

However, I did not observe any (dramatic) changes in the examples, did you run some tests? 

But they have the correct *log*-likehood term in InfoGAN :wink:  I don't know what @ppwwyyxx's plans are. But the next example I planned to move to tensorpack is InfoGAN with a distribution-class for different codes.

**Comment by Skylion007** [comment]
@PatWie I've noticed some difference on my own dataset with improved GAN. It seemed to help with my dataset in terms of yielding qualitatively better results at earlier epochs on their implementation. I maybe mistaken but isn't one important part of the paper you are missing is feature matching in section 3.1? Or have you called it something else? It also sounds like it will fix the issue you are having currently where the GAN is unstable. That sounds like possibly the most important improvement in the paper, but I could be mistaken.

**Comment by ppwwyyxx** [comment]
About improved GAN, it interests me as well because it introduces a lot of new ways of training a model, so it would be good to implement them to test whether the design of tensorpack is really flexible.

About InfoGAN, @PatWie could you elaborate what do you mean by "correct log-likelihood"? I used OpenAI code as a reference when I wrote the examples but maybe I missed something.
Also it would be nice to have distribution classes like the original code.

**Comment by PatWie** [comment]
> correct log-likelihood
I didn't saw the [sign](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/GAN/InfoGAN-mnist.py#L83).

**Comment by Skylion007** [comment]
Another issue with the wait infogan is written is that it's not easily configurable for nonmnist datasets. It would be nice to make it more flexible for other datasets. Other infogan [repos](https://github.com/JonathanRaiman/tensorflow-infogan) even let you set the discriminator and generator via the command line. Expanding infogan to run on different datasets would really also show if tensorpack is flexible. Just a thought. Using infogan on the celebA dataset for instance would be a good exercise. 

Any updates on the improved GAN? Was it included in #107 or was that just the groundwork? Looking forward to trying out the improvements.

**Comment by ppwwyyxx** [comment]
You can simply change the one line which reads the mnist dataset to two other lines which reads the CelebA dataset.
But you'll certainly need to change your architecture and it's impossible to do it automatically.
I would refuse to yse command line to define complicated things like a network, because it only makes things less flexible.

**Comment by Skylion007** [comment]
My colleague just recommended a paper on GANs: https://github.com/martinarjovsky/WassersteinGAN

Supposedly this paper uses earth mover's distance to "virtually guarantee" that GANs will converge.  Would love to see this implemented in Tensorpack. @ppwwyyxx @PatWie

**Comment by PatWie** [comment]
So they claim, they solve all issues with GAN :wink: .

As far as I can tell the differences are:
- good weight initialization
- get all weights from discriminator (alias critic), only know a hacky solution. Maybe @ppwwyyxx has an elegant way.
- add clipping (trivial)
- run several optimization steps of the discriminator (currently ugly hack)

I just came up [a few lines](https://github.com/PatWie/tensorpack-fork/tree/wgan), but I have to read the paper in detail.

**Comment by Skylion007** [comment]
So aside from the reference [implementation](https://github.com/martinarjovsky/WassersteinGAN), I found an implementation that can use [Tensorflow](https://github.com/tdeboissiere/DeepLearningImplementations/tree/master/WassersteinGAN) as backend. (It seems like it might be using a feed_dict or something because I get relatively low GPU utilization), and even one in [Chainer](http://www.itdadao.com/articles/c19a1115660p0.html).

The fact that this GAN is already implemented in so many frameworks less than a week after release bodes really well. Also I have tested at least one of the implementation and while it was unoptimized, I was impressed how quickly the loss dropped (even if the epochs did take a while).  So far it seems that the might have actually solved (most if not all) of the issues which would be really impressive!

**Comment by PatWie** [comment]
It is not my primary focus but I really like to see a working wGAN as well. So far the progress is described here #135 hopefully without ugly hacks.

**Comment by PatWie** [comment]
I think this is out-dated.

**Comment by Skylion007** [comment]
Agreed. Amazing how fast the GAN research has been in recent months. This repo has some of the best most flexible GAN implementations out there. Keep up the good work. 👍
